lampod/src/chain/blockchain.rs

```diff
@@ -27,13 +37,59 @@ impl LampoChainManager {
         LampoChainManager {
             backend: client,
             wallet_manager,
+            current_height: Arc::new(Mutex::new(Height::from_consensus(0).unwrap())),
+            best_height: Arc::new(Mutex::new(Height::from_consensus(0).unwrap())),
+            handler: RefCell::new(None),
         }
     }
 
     pub fn is_lightway(&self) -> bool {
         self.backend.is_lightway()
     }
 
+    pub fn set_handler(&self, handler: Arc<LampoHandler>) {
+        self.handler.replace(Some(handler));
+    }
+
+    pub fn handler(&self) -> Arc<LampoHandler> {
+        self.handler.borrow().clone().unwrap()
+    }
+
+    pub fn is_syncing(&self) -> bool {
+        const SYNC_THRESHOLD: u32 = 1;
+        let current_height = self.current_height.lock().unwrap();
+        let best_height = self.best_height.lock().unwrap();
+        current_height.to_consensus_u32() + SYNC_THRESHOLD < best_height.to_consensus_u32()
+    }
+
+    pub fn listen(&self) -> error::Result<()> {
+        let _backend_listner = self.backend.clone().listen()?;
+        let event_listener = self.clone();
+        std::thread::spawn(move || {
+            log::info!(target: "lampo_chain_manager", "listening for chain events");
+            loop {
+                let handler = event_listener.handler();
+                let event = handler.events().recv().unwrap();
+                match event {
+                    Event::OnChain(OnChainEvent::NewBlock(block)) => {
+                        let height = block.bip34_block_height().unwrap();
+                        log::debug!(target: "lampo_chain_manager", "new block with hash `{}` at height `{}`", block.block_hash(), height);
+                        let mut current_height = event_listener.current_height.lock().unwrap();
+                        if height as u32 > current_height.to_consensus_u32() {
+                            *current_height = Height::from_consensus(height as u32).unwrap();
+                        }
+                    }
+                    Event::OnChain(OnChainEvent::NewBestBlock((_, height))) => {
+                        log::debug!(target: "lampo_chain_manager", "new best block height `{}`", height);
+                        *event_listener.best_height.lock().unwrap() = height;
```

Comment:
thinking here, why we need a `Mutex` here? I do not think that we need and `Mutex` and still note sure there is the need of a pointer (`Arc`)